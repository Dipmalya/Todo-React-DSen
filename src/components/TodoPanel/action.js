import { DELETE_TODO } from '../../config/actionTypes';

export const deleteTodo = value => {
    return (dispatch) => {
        dispatch(deleteTodoDispatcher(value))
    };
}

const deleteTodoDispatcher = (value) => ({
    type: DELETE_TODO,
    payload: value
});