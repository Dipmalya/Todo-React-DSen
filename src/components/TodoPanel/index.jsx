import React, { Component } from "react";
import { v4 as uuidv4 } from "uuid";

const { Home: data } = CONFIG;

class TodoPanel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      list: this.props.todoList,
    };
  }

  static getDerivedStateFromProps(props) {
    const { currentView, todoList } = props;
    if (currentView === "ALL") {
      return {
        list: todoList,
      };
    } else {
      return {
        list: todoList.filter((item) => item.status === currentView),
      };
    }
  }

  onClickHandler = (event) => {
    const {
      target: { id },
    } = event;
    this.props.toggleView(id);
  };

  onRemoveHandler = (event) => {
    const {
      target: { id },
    } = event;
    this.props.deleteTodo(id);
  }

  onToggleHandler = (id) => this.props.toggleTodo(id);

  render() {
    const { currentView } = this.props;
    const { panelTypes, emptyPanel } = data;
    const { list } = this.state;

    return (
      <div className="col-md-7 offset-md-2 mt-5">
        <ul className="nav-list">
          {panelTypes.map((type) => (
            <li
              id={type}
              key={uuidv4()}
              className={type === currentView ? "selected-nav" : ""}
              onClick={this.onClickHandler}
            >
              {type}
            </li>
          ))}
        </ul>
        <div className="todo-panel">
          <ol className="todo-item">
            {list.length === 0 && (
              <div className="empty-list">
                {emptyPanel}
                <i className="ml-2 fa fa-folder-open"></i>
              </div>
            )}
            {list.map(({ id, task, status }) => (
              <li key={uuidv4()}>
                {task}
                <span className="float-right">
                  <i 
                    className={`${status === 'COMPLETED' ? 'text-success' : ''} item-icon fa fa-check-circle`}
                    onClick={this.onToggleHandler.bind(this, id)}
                  />
                  <i
                    id={id}
                    className="item-icon text-danger fa fa-trash"
                    onClick={this.onRemoveHandler}
                  />
                </span>
              </li>
            ))}
          </ol>
        </div>
      </div>
    );
  }
}

export default TodoPanel;
