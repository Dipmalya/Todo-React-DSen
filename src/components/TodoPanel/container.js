import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'
import TodoPanel from './index';

import { deleteTodo, toggleTodo } from './action';

const mapStateToProps = ({ todoList }) => ({
    todoList
});

const mapDispatchToProps = (dispatch) => ({
    deleteTodo: bindActionCreators(deleteTodo, dispatch),
    toggleTodo: bindActionCreators(toggleTodo, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoPanel);