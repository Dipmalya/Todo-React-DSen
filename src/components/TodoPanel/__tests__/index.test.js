import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import TodoPanel from '../index';

const props = {
    todoList: [
        {
            id: 'bda37a11-6a12-40ec-9365-331a516194ea',
            task: 'Do next',
            status: 'ACTIVE'
        },
        {
            id: 'bda37a11-6a12-40ec-9365-331a516194ea',
            task: 'Do nothing',
            status: 'COMPLETED'
        }
    ],
    currentView: 'ALL',
    toggleView: jest.fn(),
    deleteTodo: jest.fn(),
    toggleTodo: jest.fn()
};
const mockEvent1 = {
    target: {
        id: 'bda37a11-6a12-40ec-9365-331a516194ea'
    }
};
const mockEvent2 = {
    target: {
        id: 'ACTIVE'
    }
};

describe('TodoPanel Component', () => {
    const wrapper = mount(<TodoPanel {...props} />);

    describe('Test suite to check component rendering', () => {
        it('Test to check TodoPanel component gets rendered successfully', () => {
            expect(wrapper.exists()).toBe(true);
        });
        
        it('Test to check snapshot of the component is validated', () => {
            const component = renderer.create(<TodoPanel {...props} />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('Test to check nav list gets rendered', () => {
            const dummyNavList = wrapper.find('ul[className="nav-list"]');
            expect(dummyNavList.exists()).toBe(true);
        });

        it('Test to check number of todo nav items is 3', () => {
            const dummyNavList = wrapper.find('ul[className="nav-list"]');
            expect(dummyNavList.find('li').length).toEqual(3);
        });

        it('Test to check todo panel gets rendered', () => {
            const dummyTodoPanel = wrapper.find('div[className="todo-panel"]');
            expect(dummyTodoPanel.exists()).toBe(true);
        });

        it('Test to check number of todo items matches the props', () => {
            const dummyTodoItems = wrapper.find('ol[className="todo-item"]');
            expect(dummyTodoItems.find('li').length).toEqual(props.todoList.length);
        });

        it('Test to check action buttons on todo items gets rendered', () => {
            const dummyTodoItems = wrapper.find('ol[className="todo-item"]');
            expect(dummyTodoItems.find('i').length).toEqual(props.todoList.length * 2);
        });
    });

    describe('Test suite to check todo items is displayed as per current panel', () => {
        it('Test to check number of todo items matches the current section', () => {
            wrapper.setProps({ currentView: 'COMPLETED' });
            const completedTodos = props.todoList.filter(todo => todo.status === 'COMPLETED');
            const dummyTodoItems = wrapper.find('ol[className="todo-item"]');
            expect(dummyTodoItems.find('li').length).toEqual(completedTodos.length);
        });

        it('Test to check number of todo items matches the current section', () => {
            wrapper.setProps({ currentView: 'ACTIVE' });
            const activeTodos = props.todoList.filter(todo => todo.status === 'ACTIVE');
            const dummyTodoItems = wrapper.find('ol[className="todo-item"]');
            expect(dummyTodoItems.find('li').length).toEqual(activeTodos.length);
        });
    });

    describe('Test suite to check todo functionalities', () => {
        it('Test to check toggle todo action is called', () => {
            const mockToggleTodo = jest.fn();
            wrapper.setProps({ toggleTodo: mockToggleTodo });
            const dummyToggleLink = wrapper.find('i[className=" item-icon fa fa-check-circle"]');
            dummyToggleLink.props().onClick();
            expect(mockToggleTodo).toBeCalled();
        });

        it('Test to check delete todo action is called', () => {
            const mockDeleteTodo = jest.fn();
            wrapper.setProps({ deleteTodo: mockDeleteTodo });
            const dummyDeleteLink = wrapper.find('i[className="item-icon text-danger fa fa-trash"]');
            dummyDeleteLink.props().onClick(mockEvent1);
            expect(mockDeleteTodo).toBeCalled();
        });
    });

    describe('Test suite to check panel gets modified on changing nav list', () => {
        it('Test to check nav panel is toggled', () => {
            const mockToggleView = jest.fn();
            wrapper.setProps({ toggleView: mockToggleView });
            const dummyNavLink = wrapper.find('li[id="ACTIVE"]');
            dummyNavLink.props().onClick(mockEvent2);
            expect(mockToggleView).toBeCalled();

            const dummyTodoItems = wrapper.find('ol[className="todo-item"]');
            expect(dummyTodoItems.find('li').length).toEqual(1);
        });

        it('Test to check empty nav panel displays no list', () => {
            wrapper.setProps({ todoList: [] });
            const dummyTodoItems = wrapper.find('ol[className="todo-item"]');
            expect(dummyTodoItems.find('li').length).toEqual(0);
        });
    });
});
