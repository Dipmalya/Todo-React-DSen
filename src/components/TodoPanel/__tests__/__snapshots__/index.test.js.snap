// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPanel Component Test suite to check component rendering Test to check snapshot of the component is validated 1`] = `
<div
  className="col-md-7 offset-md-2 mt-5"
>
  <ul
    className="nav-list"
  >
    <li
      className="selected-nav"
      id="ALL"
      onClick={[Function]}
    >
      ALL
    </li>
    <li
      className=""
      id="ACTIVE"
      onClick={[Function]}
    >
      ACTIVE
    </li>
    <li
      className=""
      id="COMPLETED"
      onClick={[Function]}
    >
      COMPLETED
    </li>
  </ul>
  <div
    className="todo-panel"
  >
    <ol
      className="todo-item"
    >
      <li>
        Do next
        <span
          className="float-right"
        >
          <i
            className=" item-icon fa fa-check-circle"
            onClick={[Function]}
          />
          <i
            className="item-icon text-danger fa fa-trash"
            id="bda37a11-6a12-40ec-9365-331a516194ea"
            onClick={[Function]}
          />
        </span>
      </li>
      <li>
        Do nothing
        <span
          className="float-right"
        >
          <i
            className="text-success item-icon fa fa-check-circle"
            onClick={[Function]}
          />
          <i
            className="item-icon text-danger fa fa-trash"
            id="bda37a11-6a12-40ec-9365-331a516194ea"
            onClick={[Function]}
          />
        </span>
      </li>
    </ol>
  </div>
</div>
`;
